### PR TITLE
Custom Decay Function for Learning Rate

### PR DESCRIPTION
**Pull request by terrytangyuan**

### CODE REVIEW

File: skflow/estimators/base.py [hunk 1]
```diff
@@ -62,12 +62,15 @@ class TensorFlowEstimator(BaseEstimator):
         early_stopping_rounds: Activates early stopping if this is not None.
             Loss needs to decrease at least every every <early_stopping_rounds>
             round(s) to continue training. (default: None)
-
+        exponential_decay: Whether to apply exponential_decay to the learning_rate.
+                             A dict type containing the following keys: global_step(int),
+                             decay_steps(int), decay_rate(float), and staircase(bool).
```
Comment (ilblackdragon) [inline skflow/estimators/base.py]: what do you think about making this variable take a function?
e.g.
def learning_rate_update(learning_rate, loss, global_step):
   whatever logic for decay.
   return new_learning_rate

Comment (ilblackdragon) [inline skflow/estimators/base.py]: and then providing some default functions like exponential decay

Comment (ilblackdragon) [inline skflow/estimators/base.py]: or `learning_rate` - can be constant or a function that returns `Tensor`. Then a user can just pass `tf.train.exponential_decay` with params.

Comment (terrytangyuan) [inline skflow/estimators/base.py]: Good ideas. Consolidating them into learning_rate sounds better to me. This way there's no much change in trainer and it's much more flexible to use.

**Comment by terrytangyuan** [comment]
@ilblackdragon Changed the structure a bit and added an example.

**Comment by terrytangyuan** [comment]
@ilblackdragon Do you have any idea why this is 0.8? https://travis-ci.org/google/skflow/jobs/99976442#L2539
It's 0.86667 every time on my local pc with the seed specified and the tests passed locally. I can certainly lower the requirement to pass it on Travis but this might indicate some internal issues?

File: skflow/trainer.py [hunk 2]
```diff
@@ -45,20 +45,35 @@ class TensorFlowTrainer(object):
...
-            "learning_rate",
-            [],
-            initializer=tf.constant_initializer(learning_rate))
+        # pylint: disable=redefined-variable-type
+        if isinstance(learning_rate, float):
+            self._learning_rate = tf.get_variable(
+                "learning_rate",
+                [],
+                initializer=tf.constant_initializer(learning_rate))
+        else:
```
Comment (ilblackdragon) [inline skflow/trainer.py]: this may a bit easier to read in next way:
if isinstance(learning_rate, float):
   ...
elif callable(learning_rate):
  ...
else:
  raise ValueError(...)

Also you should add Raises: section into the __init__ doc string.

**Comment by ilblackdragon** [comment]
Sorry, didn't have time to look what's happening. Definetly we should figure out why is this happening because results should be reproducible. It's fine for now to have lowered the testing score, though.

**Comment by terrytangyuan** [comment]
No worries at all. Take your time on these. I've made the suggested changes.  :-)

**Comment by silverlining21** [comment]
@terrytangyuan  hi. 
Do you have any suggenstions for me to custom the learning rate decay? I want to implement the `inv (base_lr * (1 + gamma * iter) ^ (- power))` but have NO ideas where to start . thank you in advance.

**Comment by ilblackdragon** [comment]
You can just literally write that formula with tensorflow ops:

learning_rate = tf.pow(tf.constant(base_lr) * (1 - gamma *
tf.train.get_global_step()), -power)

Also take a look at tf.exponential_decay

In general please ask questions on Stack Overlfow or tensorflow discussion
group - there is more people looking at that.

This repo will be soon closed, as skflow is part of Tensorflow now.

**Comment by silverlining21** [comment]
OK, thanks a lot.
